ger is an alias to outer, not the other way around

albanD opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58094 Forward AD formulas batch 3
* **#59448 ger is an alias to outer, not the other way around**

Differential Revision: [D28900487](https://our.internmc.facebook.com/intern/diff/D28900487)

albanD committed:
ger is an alias to outer, not the other way around

[ghstack-poisoned]

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -571,9 +571,9 @@
-  self: grad.mv(vec2.conj())
-  vec2: grad.t().mv(self.conj())
+- name: outer(Tensor self, Tensor vec2) -> Tensor
+  self: grad.mm(vec2.unsqueeze(-1).conj()).squeeze(-1)
```
I had to modify this formula because outer supports half dtype right now but mv does not.

zou3519 replied:
We should probably add a comment and file an issue about replacing the formula with mv if/when it ever gets half support...

Are there any performance concerns here? I'm not too worried about 2 extra views here and I've rarely seen ger/outer used in the wild

albanD replied:
Actually due to more issues with cuda devices, I'll completely remove that formula and let this be autodiffed.
Note that outer was already autodiffed so that shouldn't change anything for this one.
And ger is deprecated so people should not be using it.

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28900487).

zou3519 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -9753,20 +9753,20 @@
 - func: outer(Tensor self, Tensor vec2) -> Tensor
   variants: function, method
+  dispatch:
+    CompositeExplicitAutograd: outer
```
Please fix  the comment in https://github.com/pytorch/pytorch/blob/c769300301e8bc775e54cdd114e2beaa11b1f893/torch/overrides.py#L463 as well

zou3519 replied:
(I found this by going down the list in https://github.com/pytorch/pytorch/blob/d7ef9b73fbc971c241bc3a10be4dc368d5786507/aten/src/ATen/native/native_functions.yaml#L220)

albanD replied:
Ho I didn't knew there were actual instructions for this! Good catch!

zou3519 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -9753,20 +9753,20 @@
+    CompositeExplicitAutograd: outer
 
 - func: outer.out(Tensor self, Tensor vec2, *, Tensor(a!) out) -> Tensor(a!)
+  dispatch:
+    CompositeExplicitAutograd: outer_out
```
Should this have CompositeExplicitAutograd? Comparing this to "abs.out" as an analogy, "abs.out" does not have CompositeExplicitAutograd while "abs" does.

out= functions aren't differentiable in the first place, right?

albanD replied:
They are special cased in the codegen to not be. So I don't think this changes anything. I left it for consistency with the "ger" implementation. You think I should remove it?

zou3519 replied:
I think it is better without CompositeExplicitAutograd (because it is confusing for someone who isn't familiar to the code to read through: "why does this out= function have CompositeExplicitAutograd?") but I agree that this change is "no worse than before" so we can leave it as-is

mruberry commented:
Nice fix, @albanD! Looks like @zou3519 is helping with the review, but let me know if you have a question about aliasing

albanD committed:
Update on "ger is an alias to outer, not the other way around"


Differential Revision: [D28900487](https://our.internmc.facebook.com/intern/diff/D28900487)

[ghstack-poisoned]

albanD commented:
This should be ready for final review now!

janeyx99 commented:
Reverting as this belongs to a PR that broke some forward_mode_AD tests, e.g., https://app.circleci.com/pipelines/github/pytorch/pytorch/332766/workflows/901123bb-fbd5-48ef-b654-8e6926929c8f/jobs/13988206
